### ISSUE TITLE
template tag jsdoc not evaluated with function callback as parameter

### ISSUE THREAD

**Issue by Thaina** [Microsoft/TypeScript#11724]
I made function SortFilter like this

``` js
/**
 * @param {T[]} arr
 * @template T
 */
function ItIs(arr) // for test
{
    return arr;
}

/**
 * @param {T[]} arr
 * @param {function(T):number} valuator
 * @template T
 */
function SortFilter(arr,valuator)
{
    return arr.map((item) => {
        return { item: item,value: valuator(item) };
    }).filter((pair) => Number.isFinite(pair.value)).sort((l,r) => {
        return l.value - r.value;
    }).map((pair) => pair.item);
}
```

And this is the result

![image](https://cloud.githubusercontent.com/assets/1042507/19506286/5ea4605c-95f4-11e6-8d12-bf0294134b69.png)

![image](https://cloud.githubusercontent.com/assets/1042507/19506288/647fe12c-95f4-11e6-895a-2208f3849d66.png)

As you could see. It seem like it only parse correctly when param is normal array. And fail with callback function as param

**Comment by DanielRosenwasser** [comment]
Interestingly, if you write

```ts
var x = SortFilter([1,2,3], undefined)
```

you'll get `number[]`.

Additionally, with the following

```ts
var x = SortFilter([1,2,3], q => q)
```

If you request quick info on `x` and then on `q`, `q` will correctly get the type `number` (though `x` will still have the type `T[]`.

But if you reverse the order (request quick info on `q` and then `x`), then you will get type `T[]` on `x` and `T` for `q`.

Though this might be unrelated, it'd make for a good test case.

**Comment by Thaina** [comment]
Talk about order of evaluation. I remembered this is used to work in 1.8-1.9

There is another bug I found that it may related. I like to make a chain promise with reduce

```js
var prom	= [0,1,2].reduce((promise,n) => promise.then(() => n),Promise.resolve());
```

![image](https://cloud.githubusercontent.com/assets/1042507/19512266/46a5d438-9615-11e6-8655-263588076986.png)

![image](https://cloud.githubusercontent.com/assets/1042507/19512284/57abfa5a-9615-11e6-8d7e-534c0b443d0b.png)

It seem like order of evaluation for salsa in typescript v2 was changed. So it cause problem on generic guess work
